src/lang/checkers/SanityChecker.ml

```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
+           the point in the syntax tree just before stmts.  Each item
+           in the List is List of all the accepts seen along that path
+           to this point.
+
+           Returns an updated version of the List of Lists which
+           includes any accepts seen in stmts.
+
+           Note that any given code path remains untracked until an
+           accept is discovered in it.  This keeps the seen data
+           nested List to a minimal size. *)
+        List.fold_left
+          (fun (seen2 : loc list list) (stmt : stmt_annot) ->
+            let loc = (stmt_loc stmt) in
+            match fst stmt with
+            | AcceptPayment ->
+               (match seen2 with
+                (* If we didn't see any accept statements before reaching
+                 * this point, register this as the first one. *)
+                | [] -> [[loc]]
+                (* Otherwise add this accept statement to the list of accepts
+                 * already seen on each code path reaching this point. *)
+                | _ -> List.map (fun accepts -> loc :: accepts) seen2)
+            | MatchStmt (_ident, branches) ->
+               (* For each branch in the match statement we have a
+                    new code path to "multiply" with the code paths
+                    which already reached this point, so walk each
+                    branch and build all the results into a new list.
+                *)
+               List.fold_left
+                 (fun seen3 (_pattern, branchstmts) ->
+                   match walk seen2 branchstmts with
+                   | [] -> seen3
+                   | seen4 -> seen3 @ seen4)
+                 [] branches
+            | _ -> seen2
+          ) (seen : loc list list) (stmts : stmt_annot list)
+      in
+        walk [] stmts
+    in
+
+    let check_accepts (contr : contract) =
+      let check_transition_accepts (transition : transition) =
+        let transition_accept_groups = List.map List.rev (find_accept_groups transition.tbody) in
+
+        let dup_accept_warning (group : loc list) : unit =
+          (warn1
+             (Core.sprintf "transition %s had a potential code path with duplicate accept statements:\n"
+                (get_id transition.tname) ^
+                String.concat ""
+                  (List.map
+                     (fun loc -> Core.sprintf "  Accept at %s\n" (get_loc_str loc))
+                     group))
+             warning_level_duplicate_accepts
+             (List.hd group))
+        in
+
+        List.iter
+          (fun group ->
+            match List.length group with
+            | 0 -> ()
+            | 1 -> ()
+            | _ -> dup_accept_warning group)
+          transition_accept_groups;
+
+        transition_accept_groups
+      in
+
+      let all_accept_groups =
+        (List.fold_left
+           (fun acc t -> acc @ check_transition_accepts t)
+           [] contr.ctrans)
+      in
+
+      (if List.length all_accept_groups == 0 then
```

Comment:
Same story as before - you don't want to traverse the entire list just to check whether the list is empty.

Use 

```
match all_accept_groups with
| [] -> ...
| _ -> ...
```
instead.

Reply:
Fixed.